packages/outline-view/src/browser/outline-view-tree.ts

```diff
@@ -0,0 +1,47 @@
+/********************************************************************************
+ * Copyright (C) 2019 Ericsson and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+
+import { injectable, inject, postConstruct } from 'inversify';
+import { CompositeTreeNode, TreeModelImpl, TreeSelectionService, TreeExpansionService } from '@theia/core/lib/browser';
+
+@injectable()
+export class OutlineViewTreeModel extends TreeModelImpl {
+
+    @inject(TreeSelectionService) protected readonly selectionService: TreeSelectionService;
+    @inject(TreeExpansionService) protected readonly expansionService: TreeExpansionService;
+
+    @postConstruct()
+    protected init(): void {
+        this.toDispose.push(this.tree);
+        this.toDispose.push(this.tree.onChanged(() => this.fireChanged()));
+        this.toDispose.push(this.selectionService);
+        this.toDispose.push(this.expansionService);
+        this.toDispose.push(this.expansionService.onExpansionChanged(node => {
+            this.fireChanged();
+        }));
+        this.toDispose.push(this.onOpenNodeEmitter);
+        this.toDispose.push(this.onChangedEmitter);
+        this.toDispose.push(this.treeSearch);
+    }
+
+    async collapseAll(raw?: Readonly<CompositeTreeNode>): Promise<boolean> {
```

Comment:
Why collapseAll from super implementation is not good? Does it have bugs?

Reply:
The `collapseAll` from super implementation would force a selection on the top-level parent node. In outline widget, as the selection on tree node is synchronized with the focus on the monaco editor, `collapseAll`should not re-select a different node (the tree node selection should remain the same).

Reply:
ok